File: D_star Lite/D_star Lite_demo/GeneratePath.m
function path=GeneratePath()
global s_start
global s_goal
global neighbour
global xmax
global ymax
global rhs
path=[];%存储规划路径
while ~isSamePosition(s_start,s_goal)
    temp_rhs=Inf;%临时存储rhs用于比较
    min_rhs_point=s_start;
    for i=1:size(neighbour,1)%选择s_start所有后继节点最小的g值坐标作为新的s_start
        s_temp=s_start+neighbour(i,:);
        if s_temp(1)<1||s_temp(1)>xmax||s_temp(2)<1||s_temp(2)>ymax
            continue
        end
        if temp_rhs>rhs(s_temp(1),s_temp(2))
            temp_rhs=rhs(s_temp(1),s_temp(2));
            min_rhs_point=s_temp;
        end
    end
    s_start=min_rhs_point;%选择s_start所有后继节点最小的g值坐标作为新的s_start
    if temp_rhs==Inf
        disp('No Path to GOAL!');
        break;
    else
        path=[path;s_start];%移动到s_start点，并存储到path中
    end
end

File: D star Lite C++/D_start_Lite/src/GeneratePath.cpp
/* Include files */
#include <iostream>
#include <stdio.h>

#include "main.h"
#include "Initialize.h"
#include "GeneratePath.h"
#include "isSamePosition.h"

/* Parameters Declarations */

/* Matrix Declarations */
static MatrixXd path;
static RowVector2d min_rhs_point;
static RowVector2d s_temp;
static RowVector2d s_next;

/* Function Declarations */
MatrixXd GeneratePath(void)
{
    s_next = s_start;
    path = s_start;
    while (!isSamePosition(s_next, s_goal))
    {
        //临时存储rhs用于比较
        static double min_rhs = INF;
        min_rhs_point = s_next;

        //选择s_start所有后继节点最小的rhs值坐标作为新的s_start
        for (int i = 0; i < 8; i++)
        {
            s_temp << s_next(0) + neighbour(i, 0), s_next(1) + neighbour(i, 1);
            //越界跳过
            if (s_temp(0) < 1 || s_temp(0) > xmax || s_temp(1) < 1 || s_temp(1) > ymax)
            {
                continue;
            }
            if (min_rhs > rhs(s_temp(0), s_temp(1)))
            {
                min_rhs = rhs(s_temp(0), s_temp(1));
                min_rhs_point = s_temp;

            }
        }
        s_next = min_rhs_point;
        if (min_rhs == INF)
        {
            std::cout << "No Path to Goal!" << std::endl;
            break;
        }
        else
        {
            //移动到s_next并储存进path
            appendRow(path, s_next);
        }

    }
    return path;
}